### ISSUE TITLE
Unexpected TS2430 error with exactOptionalPropertyTypes enabled

### ISSUE THREAD

**Issue by alandre** [microsoft/TypeScript#46004]
# Bug Report

### 🔎 Search Terms

- exactOptionalPropertyTypes
- Interface incorrectly extends

### 🕗 Version & Regression Information

- This is a type issue which came with v4.4.2 and `"exactOptionalPropertyTypes": true`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.2#code/PTAEBcAsFMITwA7QM6gGYCcD2BbUABHAQ3GgwEsiAbAWgFdzgBjLDaAKHIDtSM0imsAArYEyAGLYcAJWgDwoAN7tQoLFwDCVckwDWAfgBcoABQBKUAF4AfKABuWcgBMA3OwC+7TjzL9BoESwxSVwAWRIySioAVQBJJRU1TW09I1NzK1sHZwsAH1A6LidoNG5oVw8vaAAPBFYFcERYABUYZqaQnHDeKLirANEJKW7I6j6AMgGgodxZeTd2EAgYeCRQblAcOFAEbAAraCZwdhq6jAVuXj9YADkiOwAhLHBwXAAZcmRwWNI8GtIiqhWtB2khOiMKGN4spVKo7NQ6NBjF8KFwAOZuTxAA)

### 💻 Code

```ts
// the types from @material-ui/core
interface PropsFromReact {
  onClick?: () => void;
}

interface PropsFromMaterialUI {
  onClick?: (() => void) | undefined;
}

export type TheTypeFromMaterialUI = PropsFromReact & PropsFromMaterialUI;

// the type from my project
// TS2430: Interface 'NavBottomListItem' incorrectly extends interface 'TheTypeFromMaterialUI'
// Types of property 'onClick' are incompatible.
// Even if I don't extend onClick here
export interface NavBottomListItem extends TheTypeFromMaterialUI {
    value: string;
}
```

### 🙁 Actual behavior

Got error
```
Interface 'NavBottomListItem' incorrectly extends interface 'TheTypeFromMaterialUI'.
  Type 'NavBottomListItem' is not assignable to type 'PropsFromReact'.
    Types of property 'onClick' are incompatible.
      Type '((() => void) & (() => void)) | undefined' is not assignable to type '() => void'.
        Type 'undefined' is not assignable to type '() => void'.
```

### 🙂 Expected behavior

There shouldn't be any error :)

First of all, there shouldn't be an error of incorrect extending, because I don't try to override `onClick` property in `NavBottomListItem`. As I see, the type `TheTypeFromMaterialUI` is calculated to `{ onClick?: () => void; }`, so `| undefined` is just removed from the intersection. This is an understandable behavior, so normally I should be able to extend this type with `{ value: string }`

**Comment by fatcerberus** [comment]
> exactOptionsPropertyTypes

I assume you mean `exactOptionalPropertyTypes`... just asking because I've seen it *consistently* typo'd the way you have several times.

**Comment by alandre** [comment]
@fatcerberus Yes, you're right. Thanks for noticing, I've changed the description.

**Comment by ahejlsberg** [comment]
Yeah, this is definitely wrong. We're not doing the right thing when intersecting `undefined` from a missing property with an explicit `undefined` type. Internally we have different type identities for the two, and our intersection logic isn't providing for that. Here's a simple repro:

```ts
type U1 = undefined;  // Explicit undefined type
type U2 = { x?: never }['x'];  // undefined from missing property

type U3 = U1 & U2;  // never, but should be undefined
```
